File: 回溯/39.组合总和.go
func combinationSum(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	res := make([][]int, 0)
	path := make([]int, 0)
	var dfs func(int, int)
	dfs = func(sum, start int) {
		if sum == target {
			res = append(res, slices.Clone(path))
			return
		}
		if sum > target || start == len(candidates) {
			return
		}
		for i := start; i < len(candidates); i++ {
			path = append(path, candidates[i])
			sum += candidates[i]
			dfs(sum, i)
			path = path[:len(path)-1]
			sum -= candidates[i]
		}
	}
	dfs(0, 0)
	return res
}

File: 回溯/39.组合总和.java
class Solution {
    List<List<Integer>> res;
    List<Integer> path;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        res=new ArrayList<>();
        path=new ArrayList<>();
        dfs(candidates,0,0,target);
        return res;
    }

    void dfs(int[] candidates,int start,int sum,int target){
        if(sum == target){
            res.add(new ArrayList<>(path));
            return;
        }
        if(sum > target || start==candidates.length){
            return;
        }
        for(int i=start;i<candidates.length;i++){
            path.add(candidates[i]);
            sum+=candidates[i];
            dfs(candidates,i,sum,target);
            path.removeLast();
            sum-=candidates[i];
        }
    }
}